content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,586 @@
+```typescript
+const [data, setData] = createSignal<User | Admin>() // T: Signal<User | Admin | undefined>
+
+return(
+    <Show when={() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })}>
+      {user => <User user={user()} />}
+    </Show>
+  )
+```
+
+This method leverages the type information embedded in the [`<Show>`](/reference/show) component.
+TypeScript can infer that since `n` comes from a truthy `name()`, it can't be `undefined` or other falsy values.
+However, accessing this null asserted form incorrectly may thrown an error if the condition is not satisfied.
+
+## Advanced JSX attributes and directives
+
+### Custom event handlers
+
+For custom events in Solid, the [`on:___`]() and [`oncapture:___`]() attributes can be utilized. 
+To type these events, you'll need to extend Solid's JSX namespace.
+
+```typescript
+class NameEvent extends CustomEvent {
+  type: "Name";
+  detail: { name: string };
+
+  constructor(name: string) {
+    super("Name", { detail: { name } });
+  }
+}
+
+declare module "solid-js" {
+  namespace JSX {
+    interface CustomEvents {
+      Name: NameEvent;  // Matches `on:Name`
+    }
+    interface CustomCaptureEvents {
+      Name: NameEvent;  // Matches `oncapture:Name`
+    }
+  }
+}
+
+// Usage
+<div on:Name={(event) => console.log("name is", event.detail.name)} />;
+```
+
+<Aside>
+**Note**: By default, using native events like `mousemove` with the `on` prefix (e.g., <div on:mousemove={e => {}} />) will result in a TypeScript error. 
+This is because they are not included in the custom event typings for Solid.
+To solve this, extend the `CustomEvents` and `CustomCaptureEvents` interfaces to include events from the `HTMLElementEventMap`:
+
+To include all native events:
+
+```typescript
+declare module "solid-js" {
+  namespace JSX {
+    interface CustomEvents extends HTMLElementEventMap {}
+    interface CustomCaptureEvents extends HTMLElementEventMap {}
+  }
+}
+```
+
+To include specific native events, you can pick certain events (e.g. `mousemove` and `pointermove`):
+
+```typescript
+declare module "solid-js" {
+  namespace JSX {
+    interface CustomEvents extends Pick<HTMLElementEventMap, "mousemove" | "pointermove"> {}
+  }
+}
+```
+</Aside>
+
+#### Forcing properties and custom attributes 
+
+Solid's [`prop:___`]() allows forcing properties, while [`attr:___`]() permits custom attributes. 
+
+```typescript
+declare module "solid-js" {
+  namespace JSX {
+    interface ExplicitProperties {
+      count: number;
+      name: string;
+    }
+    interface ExplicitAttributes {
+      count: number;
+      name: string;
+    }
+  }
+}
+
+// Usage
+<Input prop:name={name()} prop:count={count()}/>
+<my-web-component attr:name={name()} attr:count={count()}/>
+```
+
+#### Custom directives
+
+Custom directives can be applied using the [`use:___`]() attribute.
+
+```typescript
+function model(element: HTMLInputElement, value: Accessor<Signal<string>>) {
```

Comment:
There is no way to restrict `model` directive to only input elements, so typing the params like this might lead to less issues:
```suggestion
function model(
  element: Element, // directives can be used on any HTML and SVG element
  value: Accessor<Signal<string>> // second param will always be an accessor in case value being reactive
) {
```